### ISSUE TITLE
RFC: Provide methods for sending signals natively in Rakudo/Perl 6

### ISSUE THREAD

**Issue by jstuder-gh** [rakudo/rakudo#1960]
### Current state of Signals in Rakudo:

We have a ```signal``` sub that registers signals, returning a ```Supply``` that users can tap to execute custom P6 code upon the process receiving a signal. There are 35 signal enumerations available to the P6 user (some GNU/Linux specific, some BSD specific, etc).

For signals that are available on the host system, constants are created whose Integer values match those of the host system.

For signals that are not available on the system, constants are still created with Integer values set to zero.

### The Problem:
Having unavailable signals set to zero provides [various benefits](https://github.com/rakudo/rakudo/issues/1891#issuecomment-397753723), one of which being that if the user sends a signal via the native C kill function that they do not send the wrong signal (see related issue [here](https://github.com/rakudo/rakudo/issues/1588)). However, as @jnthn [pointed out](https://github.com/rakudo/rakudo/issues/1891#issuecomment-397811793), this benefit could simulataneously be a large problem, as silent non-transmittal of an important signal could be just as bad or possibly even worse given the circumstances.

Right now, we do not offer a 'Perl6ish' way for users to send signals to arbitrary processes (that I am aware of), which may lead users to reach for the native C solution, which we cannot control as well.

With our own interface for sending signals we can filter out those that are not available on the system and alert the user.

### Proposal:
I propose that we provide an interface for handling and emitting signals through a role applied to Signal. We would provide ```handler``` and ```emit``` methods that allow users to create a sighandler for a given signal and emit signals to the specified PID, respectively.

### Additional Details:
I have an implementation (which may/probably need some additional work) showcasing what I am proposing. [Here](https://github.com/MoarVM/MoarVM/compare/master...jstuder-gh:emit_signal_op) is the MoarVM component and [here](https://github.com/rakudo/rakudo/compare/master...jstuder-gh:signal_emit_handle) is the Rakudo component.
(Note that the MoarVM code is lacking all the mapping code emitted by update_ops (omitted so as not to obscure the rest of the code) and I didn't include the NQP code that exposes the MVM op.)

Via this interface, the following example is possible:

```perl6
# Create handlers multiple sigs, and tap the returned Supply.
# Current impl wraps signal sub which would remain as-is.
my $t = Signal.handler(SIGINT, SIGEMT, SIGBREAK, SIGUSR1)
              .tap({ say "This is $_" });
LEAVE $t.close;
#`( OUTPUT

    The following signals are not supported on this system (linux): SIGEMT, SIGBREAK
    No handlers created for the signals specified above.
      in block <unit> at -e line 1
)

# Emit multiple sigs to the specified PID.
# Current impl, only one of each sig sent. But the results return, one elem per sig arg
# (with RV reused on duplicates)
my @res = Signal.emit($*PID, SIGINT, SIGINT, SIGEMT, SIGBREAK, SIGUSR1, SIGINT, SIGEMT, SIGUSR1);
#`( OUTPUT

    The following signals are not supported on this system (linux): SIGEMT, SIGBREAK
    The signals specified above have not been emitted.
      in block <unit> at -e line 1
)

say "RV of each sig passed to emit: {@res.perl}";
# OUTPUT: RV of each sig passed to emit: [0, 0, Any, Any, 0, 0, Any, 0]

sleep 2;
say "done now";
#`( OUTPUT

    This is SIGINT
    This is SIGUSR1
    done now
  )
```

Current implementation also allows calling ```handler``` and ```emit``` directly upon definite ```Signal``` instances. The below example is possible:

```perl6
my $t = SIGINT.handler.tap({ say 'I should clean up a bit...' });
LEAVE $t.close;

SIGINT.emit($*PID);
sleep 2;
say "done now";

#`( OUTPUT

    I should clean up a bit...
    done now
)
```

I think that allowing an interface for users to send signals through Rakudo would mitigate some of the issues around the native C kill function silently not sending the requested signal. We can advertise and encourage the use of the Perl 6 API as the proper way to utilize signals via P6, resulting in fewer people reaching for the native solution.

What do you think, P6devs, to providing this interface or something similar?

**Comment by jstuder-gh** [comment]
### Potential issue - Windows:

Behavior of the ```emit``` interface proposed above is different on Windows. Since the POSIX-style signals are not supported, much of the signal support on Windows is emulated via libuv. So when the the process is notified that console window's dimesions are changed for instance, libuv will emulate having received a SIGWINCH signal and execute the sighandler if one was provided by the user (see [here](http://docs.libuv.org/en/v1.x/signal.html) for the signal support libuv provides).

As far as libuv's support for sending signals to arbitrary processes, it is limited. If the signal is SIGTERM, SIGKILL, or SIGINT it will unconditionally terminate the process (with no sighandler involved). All other signals are invalid.

So, to reiterate, libuv will execute the sighandlers (as if having received a signal) upon observing certain events, but you cannot use libuv's kill interface to invoke it yourself. Because of this, the SIGINT example below will not work as the user expects:

```perl6
say $*KERNEL.name;
# OUTPUT: win32

my $t = SIGINT.handler.tap({ say 'I should clean up a bit...' });
LEAVE $t.close;

SIGINT.emit($*PID);
sleep 2;
say "done now";

# RESULT: No output. Process unconditionally terminated.
```
It is easy enough to modify libuv so that it will invoke the sighandlers if it detects that it the process is signalling itself (see change [here](https://github.com/libuv/libuv/compare/v1.19.2...jstuder-gh:uv_kill_self)).
Using that modification, the following Perl 6 code would work as the user expects:

```perl6
say $*KERNEL.name;
# OUTPUT: win32

my @sigs = Signal.^enum_value_list.grep: * !~~ SIGKILL | SIGSTOP | 0;
my $t    = Signal.handler(|@sigs).tap({ say "This is $_" });
LEAVE $t.close;

say Signal.emit($*PID, |@sigs);
sleep 2;
say 'done now';

#`( OUTPUT

    [0 0 0 0 0 0 0 0 0]
    This is SIGHUP
    This is SIGINT
    This is SIGILL
    This is SIGABRT
    This is SIGFPE
    This is SIGSEGV
    This is SIGTERM
    This is SIGWINCH
    This is SIGBREAK
    done now
)
```

However, when signalling any other process (even another process using libuv with signal handlers set up), it will just do the termination or nothing at all, which could possibly lead to some confusion due to its inconsistency. And sending signals to your own process probably isn't very useful.

Even though signalling your own process doesn't really buy us much, it requires a modification to libuv that would either need to be submitted back upstream or would need to be maintained. Maintaining a fork of libuv seems to me like it wouldn't be worth the trouble. Then there is the trouble of implementation on backends not using libuv.

At the end of the day, signals are a bit foreign to Windows applications, so there may not be an expectation that it would work anyway. Would it make sense to have ```Signal.emit``` return a ```Failure``` stating "Emit is not supported on Windows"?

**Comment by jstuder-gh** [comment]
tl;dr Should we have an interface for sending signals through Perl 6, so that users will be dissuaded from using the C kill function directly? Windows could be tricky though.

**Comment by jstuder-gh** [comment]
I'm wondering if anyone has any opinion on this functionality?

Personally I think it would be nice to have something like this, and I have a preliminary implementation on MoarVM ready (included in first post on thread).

**Comment by lizmat** [comment]
Closing this now.  Please re-open as a problem solving issue referring to this closed issue if you disagree.
